bot/utils/redis_cache.py

```diff
@@ -0,0 +1,360 @@
+    """
+
+    _namespaces = []
+
+    def __init__(self) -> None:
+        """Initialize the RedisCache."""
+        self._namespace = None
+        self.bot = None
+        self.increment_lock = asyncio.Lock()
+
+    def _set_namespace(self, namespace: str) -> None:
+        """Try to set the namespace, but do not permit collisions."""
+        # We need a unique namespace, to prevent collisions. This loop
+        # will try appending underscores to the end of the namespace until
+        # it finds one that is unique.
+        #
+        # For example, if `john` and `john_`  are both taken, the namespace will
+        # be `john__` at the end of this loop.
+        while namespace in self._namespaces:
+            namespace += "_"
+
+        log.trace(f"RedisCache setting namespace to {self._namespace}")
+        self._namespaces.append(namespace)
+        self._namespace = namespace
+
+    @staticmethod
+    def _to_typestring(value: RedisType) -> str:
+        """Turn a valid Redis type into a typestring."""
+        for prefix, _type in TYPESTRING_PREFIXES:
+            if isinstance(value, _type):
+                return f"{prefix}{value}"
+        raise TypeError(f"RedisCache._from_typestring only supports the types {NICE_TYPE_LIST}.")
+
+    @staticmethod
+    def _from_typestring(value: Union[bytes, str]) -> RedisType:
+        """Turn a typestring into a valid Redis type."""
+        # Stuff that comes out of Redis will be bytestrings, so let's decode those.
+        if isinstance(value, bytes):
+            value = value.decode('utf-8')
+
+        # Now we convert our unicode string back into the type it originally was.
+        for prefix, _type in TYPESTRING_PREFIXES:
+            if value.startswith(prefix):
+                return _type(value[len(prefix):])
+        raise TypeError(f"RedisCache._to_typestring only supports the prefixes {NICE_PREFIX_LIST}.")
+
+    def _dict_from_typestring(self, dictionary: Dict) -> Dict:
+        """Turns all contents of a dict into valid Redis types."""
+        return {self._from_typestring(key): self._from_typestring(value) for key, value in dictionary.items()}
+
+    def _dict_to_typestring(self, dictionary: Dict) -> Dict:
+        """Turns all contents of a dict into typestrings."""
+        return {self._to_typestring(key): self._to_typestring(value) for key, value in dictionary.items()}
+
+    async def _validate_cache(self) -> None:
+        """Validate that the RedisCache is ready to be used."""
+        if self.bot is None:
+            log.exception("Attempt to use RedisCache with no `Bot` instance.")
+            raise RuntimeError(
+                "Critical error: RedisCache has no `Bot` instance. "
+                "This happens when the class RedisCache was created in doesn't "
+                "have a Bot instance. Please make sure that you're instantiating "
+                "the RedisCache inside a class that has a Bot instance "
+                "class attribute."
```

Comment:
Let's drop "class" cause it may imply to some that instance attributes aren't acceptable.

```suggestion
                "attribute."
```